### ISSUE TITLE
Rule Request: Acronyms Camel Case

### ISSUE THREAD

**Issue by moskvin** [realm/SwiftLint#1850]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Rule Request

1. Why should this rule be added? Share links to existing discussion about what
   the community thinks about this.

https://google.github.io/styleguide/javaguide.html#s5.3-camel-case
https://stackoverflow.com/a/27172000

2. Provide several examples of what _would_ and _wouldn't_ trigger violations.

would:
```swift
let userID = 1
let greatAPI
```
wouldn't
```swift
let userId = 1
let greatApi
```
3. Should the rule be configurable, if so what parameters should be configurable?

no, we need just enable/disable

4. Should the rule be opt-in or enabled by default? Why?

no

**Comment by SDGGiesbrecht** [comment]
> Should the rule be configurable [...]?
> no, we need just enable/disable.

Well, actually...

Those links are from Google’s Java style guide and a discussion of Microsoft’s .NET style guide.

Swift’s own [API Design Guidelines](https://swift.org/documentation/api-design-guidelines/#follow-case-conventions) directly contradict them, and are enforced for the APIs of the Swift Standard Library, Foundation, etc.

According to this standard, the rule would instead behave thus:
```swift
var utf8Bytes: [Utf8.CodeUnit] // ✗
var utf8Bytes: [UTF8.CodeUnit] // ✓

var isRepresentableAsAscii = true // ✗
var isRepresentableAsASCII = true // ✓

var userSmtpServer: SecureSmtpServer // ✗
var userSMTPServer: SecureSMTPServer // ✓
```

So yes, apparently such a rule would need to be configurable. ;)

***

Practically speaking though, do you have a suggestion for how to differentiate between what is and is not an acronym? The only thing I can think of is checking against a list, and I do not imagine that being very reliable.

However, I like the idea of such a rule if a way can be found to do it well. Спасибо/thank you for your suggestion, @moskvin.

**Comment by moskvin** [comment]
@SDGGiesbrecht thank you for additional info for the rule. Yes, you are right we have exclusions for some abbreviations and my suggestion to configure it for teams individually.

**Comment by marcelofabri** [comment]
I think those could be two different rules. This one would be opt-in and the other one (described by @SDGGiesbrecht) would be enabled by default.

This one is probably easier to think about, as you can say that 2+ uppercase characters in sequence are an acronym (although I guess this might not be true always).

Enforcing to make those uppercase is harder though: we'd probably need to have a list and to figure out what is a "word" in an identifier, which might be tricky.

**Comment by craig-miller** [comment]
Even more fun is that Swift enums should begin with a lowercase letter.  So if you have enums for JPEG, GIF, etc is it JPEG or jpeg in an enum?

**Comment by SDGGiesbrecht** [comment]
@craig-miller
> ...is it JPEG or jpeg in an enum?

[Apple/Swift style](https://swift.org/documentation/api-design-guidelines/#follow-case-conventions):
```swift
enum ImageType {
    case jpeg
    // or
    case jointPhotographicExpertsGroup
}
```

[Google/Java style](https://google.github.io/styleguide/javaguide.html#s5.2.4-constant-names) (discussed in earlier comments):
```java
enum ImageType {
    JPEG,
    // or
    JOINT_PHOTOGRAPHIC_EXPERTS_GROUP
}
```

[Microsoft/.NET style](https://msdn.microsoft.com/en-us/library/141e06ef(v=vs.71).aspx) (discussed in earlier comments):
```vba
Enum ImageType
    Jpeg
    ' or
    JointPhotographicExpertsGroup
End Enum ' ImageType
```

**Comment by craig-miller** [comment]
@SDGGiesbrecht It's ambiguous.  The latest from Apple is that enums should begin with a lowercase letter AND that acronyms should be uppercased.  So .jpeg in an enum and JPEG everywhere else (Method names, class names, etc)?

**Comment by SDGGiesbrecht** [comment]
@craig-miller, if you follow the above link to the *API Design Guidelines* at Swift.org, and then click on “More Detail”, you will see:

> **Follow case conventions**. Names of types and protocols are `UpperCamelCase`. Everything else is `lowerCamelCase`.
>
> ▼ LESS DETAIL
>
> Acronyms and initialisms that commonly appear as all upper case in American English should be uniformly up‐ or down‐cased according to case conventions:
>
> ```swift
> var utf8Bytes: [UTF8.CodeUnit]
> var isRepresentableAsASCII = true
> var userSMTPServer: SecureSMTPServer
> ```

The definitive example is the first one: `var utf8Bytes`

**Comment by SDGGiesbrecht** [comment]
@craig-miller,
> So .jpeg in an enum and JPEG everywhere else (Method names, class names, etc)?

So `jpeg` when a normal word would not be capitalized:
- At the beginning of variable, function, or method name.

And `JPEG` where a normal word would be capitalized:
- In the middle or at the end of any identifier.
- At the beginning of a class or protocol name.

**Comment by craig-miller** [comment]
Yep, that's how I'm understanding things.
